src/rpc/mining.cpp

```diff
@@ -482,9 +482,8 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
         // TODO: Maybe recheck connections/IBD and (if something wrong) send an expires-immediately template to stop miners?
     }
 
-    const struct VBDeploymentInfo& segwit_info = VersionBitsDeploymentInfo[Consensus::DEPLOYMENT_SEGWIT];
     // GBT must be called with 'segwit' set in the rules
-    if (setClientRules.count(segwit_info.name) != 1) {
+    if (setClientRules.count("segwit") != 1) {
```

Comment:
I don't think this is sufficient -- getblocktemplate only populates `aRules` based on versionbits state so whether or not "segwit" is passed as an arg, it ends up outputting an empty array `{"rules": []}` for the template. I think that conflicts with [BIP145](https://github.com/bitcoin/bips/blob/master/bip-0145.mediawiki): "Transactions with witness data may only be included if the template's "rules" list (see BIP 9) includes "segwit"."

Reply:
I think we're not compliant with this part of BIP 145 anyway, specifically:

> The '!' rule prefix MUST be enabled on the "segwit" rule for templates including transactions with witness data.

There's no code to insert the '!' in templates with blocks containing segwit transactions. I've also tested on master that `getblocktemplate` doesn't include `!segwit` as a rule for templates with witness data.

The fact that this has always been broken suggests to me that no-one is looking at the `rules` field on the return object.

I can add some code to update the `rules` based on whether csv and segwit are activated, but it doesn't seem worth it.

Reply:
I don't think it makes much sense to require `segwit` in the params if we're not going to bother reporting it ourselves. I think the additional code would just be:

```
    if (pindex->nHeight+1 >= consensusparams.CSVHeight) aRules.push_back("csv");
    if (!fPreSegWit) aRules.push_back("segwit");
```

(also, could have said `fPreSegWit = !IsWitnessEnabled(pindexPrev, consensusParams);`)

Looks like including the `!` got broken prior to activation in March 2017 in #9955 with segwit's `gbt_force` value changing from false to true, so bitcoind hasn't been compliant with bip145 since 0.15, though 0.14 should have been.

I think the logic here isn't right and we should have two settings in `VBDeploymentInfo`, `gbt_simplified` which just adds the `!` when false, and `gbt_optional` which, when the client doesn't specify the rule, avoids setting the versionbit while signalling if `gbt_optional` is true, and errors out when the rule is activated and `gbt_optional` is false. segwit should initially have had `gbt_simplified=false` and `gbt_optional=true`, and just changed `gbt_optional=false` in #14811.

There's no provision in bip9 or bip145 for ever removing forks from the rules list, which might be worth improving, as an alternative.

@luke-jr -- thoughts? You added the gbt changes for bip9 and acked 9955 :)

Reply:
I can't see how #9955 changed including the `!`. Am I missing something?

Reply:
ah, I see: https://github.com/bitcoin/bitcoin/pull/9955/files#diff-d6954440f0346f8e42dcd669dd4aeafaL20

'force' is defined as "Whether GBT clients can safely ignore this rule in simplified usage" which seems backwards to me.

Reply:
I guess `gbt_force` is too unclear a parameter name...

We should fix this. Maybe in a separate PR.